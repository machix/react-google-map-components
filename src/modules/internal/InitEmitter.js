import noop from "lodash/noop";
import pull from "lodash/pull";
import isFunction from "lodash/isFunction";

export const createInitEmitter = () => {
  let initArgs = null;

  const queue = [];

  return { init, onInit };

  function init(...args) {
    if (initArgs !== null) {
      throw new Error("Emitter is already initialized");
    }

    initArgs = args;

    while (queue.length > 0) {
      const fn = queue.shift();

      fn(...args);
    }
  }

  function onInit(fn) {
    if (!isFunction(fn)) {
      throw new Error("Expected init listener to be a function.");
    }

    if (initArgs) {
      fn(...initArgs);

      return noop;
    }

    queue.push(fn);

    return () => {
      pull(queue, fn);
    };
  }
};
