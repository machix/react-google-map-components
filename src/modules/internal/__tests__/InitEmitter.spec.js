import { createInitEmitter } from "../InitEmitter";

describe("InitEmitter", () => {
  it("should pass init arguments to init listeners", () => {
    const emitter = createInitEmitter();
    const fn = jest.fn();

    emitter.onInit(fn);
    emitter.init("a", "b", "c");

    expect(fn).toHaveBeenCalledTimes(1);
    expect(fn).toHaveBeenLastCalledWith("a", "b", "c");
  });

  it("should immediately call listener if emitter is initialized", () => {
    const emitter = createInitEmitter();
    const fn = jest.fn();

    emitter.init("a", "b", "c");
    emitter.onInit(fn);

    expect(fn).toHaveBeenCalledTimes(1);
    expect(fn).toHaveBeenLastCalledWith("a", "b", "c");
  });

  it("should throw on second initialize", () => {
    const emitter = createInitEmitter();

    expect(() => emitter.init()).not.toThrow();
    expect(() => emitter.init()).toThrowErrorMatchingSnapshot();
  });

  it("should return unsubscribe function", () => {
    const emitter = createInitEmitter();
    const fn = jest.fn();

    const unsubsctibe = emitter.onInit(fn);

    expect(fn).toHaveBeenCalledTimes(0);

    unsubsctibe();

    expect(fn).toHaveBeenCalledTimes(0);

    emitter.init();

    expect(fn).toHaveBeenCalledTimes(0);
  });

  it("should when listener is not a function", () => {
    const emitter = createInitEmitter();

    expect(() => emitter.onInit()).toThrowErrorMatchingSnapshot();
    expect(() => emitter.onInit(null)).toThrowErrorMatchingSnapshot();
  });
});
