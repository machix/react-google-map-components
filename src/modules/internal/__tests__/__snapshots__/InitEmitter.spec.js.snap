// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InitEmitter should throw on second initialize 1`] = `"Emitter is already initialized"`;

exports[`InitEmitter should when listener is not a function 1`] = `"Expected init listener to be a function."`;

exports[`InitEmitter should when listener is not a function 2`] = `"Expected init listener to be a function."`;
